pkg/endpoint/endpoint.go

```diff
@@ -1057,6 +1064,42 @@ func (e *Endpoint) SetK8sNamespace(name string) {
 	e.unlock()
 }
 
+// SetNamedPorts sets the k8s container ports specified by kubernetes.
+// Note that once put in place, the new k8sPorts is never changed,
+// so that the map can be used concurrently without keeping locks.
+// Reading the 'e.k8sPorts' member (the "map pointer") *itself* requires the endpoint lock!
+// Can't really error out as that might break backwards compatibility.
+func (e *Endpoint) SetNamedPorts(containerPorts []types.ContainerPort) error {
+	k8sPorts := make(policy.NamedPortsMap, len(containerPorts))
+	for _, cp := range containerPorts {
+		if cp.Name == "" {
+			continue // skip unnamed ports
+		}
+		if !api.IsNamedPort(cp.Name) {
+			log.WithField(logfields.PortName, cp.Name).Warning("ContainerPort: Invalid port name, not using as a named port")
+			continue
+		}
+		name := strings.ToLower(cp.Name)
+		u8p, err := u8proto.ParseProtocol(cp.Protocol)
+		if err != nil {
+			log.WithError(err).WithField(logfields.Protocol, cp.Protocol).Warning("ContainerPort: invalid protocol")
+			continue
+		}
+		if cp.ContainerPort < 1 || cp.ContainerPort > 65535 {
+			log.WithField(logfields.Port, cp.ContainerPort).Warning("ContainerPort: Port number out of 16-bit range")
+			continue
+		}
+		k8sPorts[name] = policy.NamedPort{
```

Comment:
Should we warn on duplicates?

Reply:
This is for the Endpoint, so these come from the POD spec, and I'd expect K8s to not allow duplicate names in the POD spec?

Reply:
Not sure. I would warn so we at least know.